saw-core/prelude/Prelude.sawcore

```diff
@@ -2989,10 +2928,88 @@ emptyFunStack = Nil1 (List1 LetRecType);
 pushFunStack : List1 LetRecType -> FunStack -> FunStack;
 pushFunStack frame stack = Cons1 (List1 LetRecType) frame stack;
 
+-- The type of FunStackE
+FunStackE_type : (E:EvType) -> FunStack -> sort 0;
+FunStackE_type E stack =
+  List1#rec
+  (List1 LetRecType)
+  (\ (_:FunStack) -> sort 0)
+  (Either String (evTypeType E))
+  (\ (frame:List1 LetRecType) -> \ (_:FunStack) -> \ (E':sort 0) ->
+     Either (FrameCall frame) E')
+  stack;
+
+-- The encoding of FunStackE
+FunStackE_enc : (E:EvType) -> (stack:FunStack) -> FunStackE_type E stack -> sort 0;
+FunStackE_enc E stack =
+  List1#rec
+  (List1 LetRecType)
+  (\ (stack:FunStack) -> FunStackE_type E stack -> sort 0)
+  (\ (e:Either String (evTypeType E)) ->
+     Either#rec String (evTypeType E) (\ (_:Either String (evTypeType E)) -> sort 0)
+       (\ (_:String) -> Void) (evRetType E) e)
+  (\ (frame:List1 LetRecType) -> \ (stack:FunStack) -> \ (rec:FunStackE_type E stack -> sort 0) ->
+   \ (e:Either (FrameCall frame) (FunStackE_type E stack)) ->
+     Either#rec (FrameCall frame) (FunStackE_type E stack) (\ (_:Either (FrameCall frame) (FunStackE_type E stack)) -> sort 0)
+       (FrameCallRet frame) rec e)
+  stack;
+
+-- The event type corresponding to a FunStack
+FunStackE : (E:EvType) -> FunStack -> EvType;
+FunStackE E stack = Build_EvType (FunStackE_type E stack) (FunStackE_enc E stack);
+
+
 -- The monad for specifications (FIXME: document this!)
 primitive SpecM : (E:EvType) -> FunStack -> sort 0 -> sort 0;
 
 
+-- The type of the preconditions needed for refinesS
```

Comment:
Explain in the comments that this is the type of a relation between `FunStackE` objects